### ISSUE TITLE
warning CA1308 : replace the call to 'ToLowerInvariant' with 'ToUpperInvariant' when I am obliged to return lower case string.

### ISSUE THREAD

**Issue by TanvirArjel** [dotnet/roslyn-analyzers#2904]
It make sense to use 'ToUpperInvariant' when comparing two string but when I am obliged to return a lower case string as follows the warning does not make sense:

```
private static string FormatKey(string unformattedKey)
{
            var result = new StringBuilder();
            int currentPosition = 0;
            while (currentPosition + 4 < unformattedKey.Length)
            {
                result.Append(unformattedKey.Substring(currentPosition, 4)).Append(" ");
                currentPosition += 4;
            }
            if (currentPosition < unformattedKey.Length)
            {
                result.Append(unformattedKey.Substring(currentPosition));
            }

            return result.ToString().ToLowerInvariant(); // This line showing the warning
  }
```

I think this warning should be removed in case of such method's return value. Any idea please!

**Comment by mavasani** [comment]
@TanvirArjel It seems very application specific when such an exception to the rule is required and should be applied. Note that analyzers are designed to be generic in nature, and lot of these analyzers might flag violations which are valuable for some code base, but false positives for others. You should add a source suppression for this case, with the suppression comment acting as a documentation on why you really need to use the `ToLowerInvariant` at that place. This also helps future developers on your code base know that they should not remove the `ToLowerInvariant` invocation at explicitly suppressed locations.

**Comment by mavasani** [comment]
Please re-open issue if you any further questions. Thanks!

**Comment by TanvirArjel** [comment]
@mavasani 

> >Note that analyzers are designed to be generic in nature

I think what I have proposed is a generic case.  if the `ToLoweInvariant` is used on the method return value analyzers should ignore this. Isn't it?

Can you give a reason why should I call `ToUpperInvariant()` when I am returning a string, while not comparing?

**Comment by TanvirArjel** [comment]
@mavasani You should not close until discussion is ended! isn't it?

**Comment by mavasani** [comment]
> I think what I have proposed is a generic case. if the ToLoweInvariant is used on the method return value analyzers should ignore this. Isn't it?

I might as well be writing:

```csharp
public static bool Equals(string str1, string str2)
  => NormalizeStringForCompare(str1) == NormalizeStringForCompare(str2);
private string NormalizeStringForCompare(string str)
   => str.ToUpperInvariant();
```

Implementing your suggestion would lead to a valid violation being masked out, which is not desireable.

> You should not close the issue until discussion is ended! isn't it?

This bucket of questions is repeatedly raised on this repo, where customers want analyzers to be tuned to match their requirements. It is an agreed upon design principle for analyzers not to special case behavior, and the resolution was following this well established principle.

**Comment by TanvirArjel** [comment]
@mavasani Okay! Got it. Thank you.
